Итоговый пул

Blackdeer1524 opened the pull request.

Blackdeer1524 committed:
init

Blackdeer1524 committed:
wip

Blackdeer1524 committed:
please send help

Blackdeer1524 committed:
basic tests

Blackdeer1524 committed:
star test

Blackdeer1524 committed:
more tests

Blackdeer1524 committed:
strong star normal form achieved?

Blackdeer1524 committed:
recursive star simplification (SSNF)

Blackdeer1524 committed:
added huge test for SSNF

Blackdeer1524 committed:
fixed test: it didn't actually asserted anything

Blackdeer1524 committed:
fixed greedy star

Blackdeer1524 committed:
more efficient alternative expansion

Blackdeer1524 committed:
minor fix

Blackdeer1524 committed:
"Давай по новой, Миша, всё - фигня"

Blackdeer1524 committed:
new parser sketch

Blackdeer1524 committed:
started reworking unary processing

Blackdeer1524 committed:
some basic tests

Blackdeer1524 committed:
fixed broken test

Blackdeer1524 committed:
greedy star fix

Blackdeer1524 committed:
all tests are passing. SSNF WIP

Blackdeer1524 committed:
trailing concat in unary test

Blackdeer1524 committed:
file rename

Blackdeer1524 committed:
checkpoint

Blackdeer1524 committed:
WIP

Blackdeer1524 committed:
ssnf done?

Blackdeer1524 committed:
integrated ssnf in basic tests

Blackdeer1524 committed:
parser fix &&  test

Blackdeer1524 committed:
star simplification

Blackdeer1524 committed:
ToString trait

Blackdeer1524 committed:
checkpoint

Blackdeer1524 committed:
alternative args sort

Blackdeer1524 committed:
minor

Blackdeer1524 committed:
removed repetitions in ssnf

Blackdeer1524 committed:
fixed AltArg type

Blackdeer1524 committed:
nested alt test

Blackdeer1524 committed:
uncommented assertion in 'star_simplification' test 💀

viacheslavek committed:
Пока что все выглядит так, тут:
- генератор
- парсер
- скелет для глушкова
- визуализация автомата
- визуализация синтаксического дерева разбора
P.S. более подробные комментарии есть у меня в репозитории

Blackdeer1524 committed:
yet another complete rewrite

Blackdeer1524 committed:
basic tests

Blackdeer1524 committed:
another test

Blackdeer1524 committed:
parsing rework

Blackdeer1524 committed:
ssnf

Blackdeer1524 committed:
removed commented blocks

Blackdeer1524 committed:
alternative's args sort

Blackdeer1524 committed:
ACI checkpoint

Blackdeer1524 committed:
fix: checkpoint

Blackdeer1524 committed:
distribution logic fix

Blackdeer1524 committed:
checkpoint

Blackdeer1524 committed:
fixed tests on simplification

Blackdeer1524 committed:
more tests on simplification

Blackdeer1524 committed:
proper handling of full simplification

Blackdeer1524 committed:
derivatives checkpoint

Blackdeer1524 committed:
elementary differentiation tests

Blackdeer1524 committed:
fixed parser flattenning

Blackdeer1524 committed:
fixed derivative (because why not)

Blackdeer1524 committed:
Done

viacheslavek committed:
Добавил генерацию слов

viacheslavek committed:
Добавил скелет для бенчмарков

Blackdeer1524 committed:
Merge pull request #3 from Blackdeer1524/lab2-lokshin

Система тестирования

Blackdeer1524 committed:
removed redundant comments

Blackdeer1524 committed:
newline char trimming in input

Blackdeer1524 committed:
Merge pull request #2 from Blackdeer1524/lab2-slesarev

Lab2 slesarev

viacheslavek committed:
Добавил связь с rust

viacheslavek committed:
Добавил запуск питоновского скрипта для сравнения выражений
- почти готово

viacheslavek committed:
Добавил не очень красивое логирование
- произвел замеры - Данил сделал отличный алгоритм, у меня match с его регуляркой работает быстрее!

viacheslavek committed:
Добавил ввод параметров из консоли

viacheslavek committed:
Merge pull request #5 from Blackdeer1524/lab2-lokshin

Связи с языками

Blackdeer1524 commented:
@TonitaN, я в итоге не выпиливал производную потому что слишком жалко :(

Blackdeer1524 committed:
chore(python benches): changed pattern + added regex compilation

Blackdeer1524 committed:
feat(go): command line args

Blackdeer1524 committed:
feat(README): basic info

Blackdeer1524 committed:
chore(go): renamed args

Blackdeer1524 committed:
feat(go/benches): better benches

Blackdeer1524 committed:
docs(README): python notice

TonitaN commented on golang/internal/benchmark/benchmark.go:
```diff
@@ -0,0 +1,159 @@
+
+	fmt.Printf(
+		"\tto before: regex: %s, status: %s, duration: %s\n",
+		wordsWithRegex.RegexBefore, okBefore, durBefore,
```
Golang, как минимум моя версия, довольно творчески выводит сообщения о времени, если не задать формат. Лучше добавить `Milliseconds` и выводить это как число.

TonitaN commented on golang/internal/benchmark/benchmark.go:
```diff
@@ -0,0 +1,159 @@
+	}
+
+	go func() {
+		<-time.After(time.Duration(len(words)) * time.Second)
```
Хотелось бы понять, зачем конвертировать в Duration длину тестирующего слова. Кажется, здесь я перестаю ориентироваться в коде.

TonitaN commented on golang/internal/gluskov/tree_to_machine.go:
```diff
@@ -0,0 +1,146 @@
+
+func (m *Machine) GetRuneBetweenStates(left, right State) (rune, State) {
+	if right == 0 {
+		return 'я', 0
```
Это был эпсилон-переход? Очевидно, где-то проблема в конструкции: в Глушкове их не бывает. Из-за этого может обрушиться тестирование, ведь такие псевдо-эпсилоны попадут в тестирующую строку.

TonitaN commented on golang/internal/gluskov/tree_to_machine.go:
```diff
@@ -0,0 +1,146 @@
+
+func (m *Machine) handleConcat(currentState State, node *syntax.Regexp, isFinal bool) []State {
+
+	leftState := m.handleRegex(node.Sub[0], currentState, false)
```
Почему финальность False? А если конкатенируемся со звездой?

TonitaN commented on golang/internal/reggen/generator.go:
```diff
@@ -0,0 +1,123 @@
+		right := rgp.generateRegexRecursive(maxLength/2, maxStarHeight)
+		return concat(left, right)
+	case 1:
+		left := rgp.generateRegexRecursive(maxLength-1, maxStarHeight)
```
Вообще под длиной имелась в виду т.н. alphabetic width регулярок - совокупное число вхождений букв в регулярку. Но народ подходит к этому вопросу творчески, и здесь я вижу задание "длины максимального пути", причём примерно, из-за особенностей округления. Ну вот собственное понимание, что такое длина регулярки - в принципе, не страшно, округление - это мелкий баг, но не принципиальный в общем.  Равномерность длин аргументов конкатенации - баг чуть погрубее, с ним бы разобраться.

TonitaN commented on golang/internal/wordgen/word_generator.go:
```diff
@@ -0,0 +1,171 @@
+
+	for i := 0; i < len(rwws); i++ {
+		fmt.Println("generating a word for regex")
+		rww, err := GenerateWordsForRegex(regexes[i], countWords, maxDumpSize)
```
Хорошо бы логировать порождаемые слова в форме `x1 (x2)^n Z`, где `x2` --- область накачки, тогда бы, в том числе, стало видно псевдо-эпсилоны в порождаемых словах.

TonitaN commented on golang/internal/gluskov/tree_to_machine.go:
```diff
@@ -0,0 +1,146 @@
+		return m.handleStar(currentState, node, isFinal)
+	case syntax.OpCapture:
+		return m.handleCapture(currentState, node, isFinal)
+	case syntax.OpCharClass:
```
А вот это - насколько мне понятно, наследие импортируемого парсера, и не нужно было его обрабатывать - это ведь знак, что что-то пошло не так в регулярке.

TonitaN commented on rust/src/aci.rs:
```diff
@@ -0,0 +1,266 @@
+                            break;
+                        }
+                        let item_repr = first_args[head_trim_number].to_string();
+                        let can_distribute = alt_args.iter().all(|item| match item {
```
Дистрибутивность же не обязана выносить общий множитель из всех скобок.
Может быть и так:
```
(a*a|b*|a*c) \\тут дистрибутивность у двух аргументов
```

TonitaN commented on rust/src/parser.rs:
```diff
@@ -0,0 +1,826 @@
+        self.next_char = None;
+        let res = self.expect_alternative();
+        self.expr_iter = None;
+        let after_ssnf = ssnf(res);
```
Почему народ помещает именно в парсер весь главный функционал? Вы не первые)

TonitaN commented on rust/src/aci.rs:
```diff
@@ -0,0 +1,266 @@
+use super::parser::ParsingResult;
+
+pub fn simplify(arg: ParsingResult) -> ParsingResult {
+    match arg {
```
Если честно, то потерялась в этой функции. И непонятно, где именно нас настигла проблема: регулярка `(a|b*)` и подобные ей преобразуются некорректно. Я подозреваю, что дело всё же не в парсере, а в упрощениях, хотя не исключаю ничего.

TonitaN reviewed (changes requested):
Из критичного:
1. автомат Глушкова строится некорректно (лучше ещё раз просчитать алгоритм, чтобы результат совпадал с тем, который через First/Follow/Last)
2. потерялась проверка на эквивалентность исходной и трансформированной регулярки в модуле тестирования. Т.к. сейчас тестируются только заведомо не подходящие по словарю строки (нагрузочно, с накачкой) и результат их разбора очевидно будет у исходной и преобразованной регулярок одинаковым. Перед этим стоило проверить корректность преобразования на просто случайных путях в автомате Глушкова.
3. где-то в части Данилы регулярки меняют семантику. Пример: `(a|b*)`. Нужно разобраться с этим вопросом и протестировать после изменений на случайных регексах.

Если честно, часть Данилы уже смотрела не так внимательно (функция simplify написана довольно запутанно). В качестве благого пожелания - написать к ней комментарии. Все остальные мои заметки - тоже "для души". Не хочется выносить частичную дистрибутивность из альтернатив - ну и не надо, зачёт будет и без этого. Аналогично - с корректировкой порождения случайных регулярок в модуле Вячеслава.

Blackdeer1524 committed:
fixed corners

Blackdeer1524 committed:
feat(rust): added test that covers first star in alternative

Blackdeer1524 committed:
Merge pull request #8 from Blackdeer1524/lab2-slesarev

Lab2 slesarev

TonitaN commented:
@Blackdeer1524 , теперь всё корректно, спасибо! Ждём исправления Славы.

viacheslavek committed:
Поправил ревью часть 1
- в звезде Клини
- конкатенации
- добавил эквивалентность

viacheslavek committed:
Поправил ревью часть 2
- обработка финальных состояний
- конкатенация в дереве
- генерация слов для эквивалентности

TonitaN commented on golang/internal/benchmark/benchmark.go:
```diff
@@ -0,0 +1,212 @@
+	reg *reggen.Regexes, rustBinaryPath string, countWords, maxDumpSize int,
+) ([]wordgen.RegexesWithWords, error) {
+
+	regexes := reg.Generate()
```
В итоге получается два разных набора регулярок в разных системах тестов.

TonitaN commented on golang/internal/reggen/generator.go:
```diff
@@ -0,0 +1,123 @@
+}
+
+func (rgp *RegexGeneratorParams) Generate() string {
+	regex := rgp.generateRegexRecursive(rgp.letterCount, rgp.alphabetSize)
```
😥

TonitaN commented on golang/internal/reggen/generator.go:
```diff
@@ -0,0 +1,123 @@
+		return letter
+	default:
+		subExpr := rgp.generateRegexRecursive(maxLength-1, maxStarHeight-1)
+		return starKellie(subExpr)
```
Параметр звёздная высота - да кому я тут нужен?

TonitaN reviewed (changes requested):
Эпсилоны в тестовых строках... :(
Простим баги генератора (генераторы почти никому не удались), и странное поведение парсера регулярок у Go. Но пока в тестовых строках есть явные эпсилоны, слишком очевидно, что с тестированием дело неладно.

viacheslavek committed:
Поправил ревью часть 3
- баги в поиске циклов
- баги в генераторе
- заменил логер
- убрал лишний вывод
- теперь регулярки генерируются один раз

viacheslavek committed:
Увеличил число слов для эквивалентности

viacheslavek committed:
Поправил форматирование времени

TonitaN reviewed (approved):
Выглядит адекватно. Зачту.
